stdlib/public/Concurrency/TaskPrivate.h

```diff
@@ -167,36 +169,123 @@ class TaskFutureWaitAsyncContext : public AsyncContext {
+/// * The locked state of the status record
+/// * The current maximum priority of the task
+/// * The identity of the thread executing the task
+/// * Pointer to head of TaskStatusRecords
+///
+/// It is important for all of this information to be in the same atomic so that
+/// when the task's state changes, the information is visible to all threads that
+/// may be modifying the task, allowing the algorithm to eventually converge.
+///
+/// Some changes to task state, like cancellation are local to the concurrency
+/// runtime library but others like priority escalation require deeper integration
+/// with the OS in order to have the intended side effects. On Darwin, Swift
+/// Concurrency Tasks runs on dispatch's queues. As such, we need to use an
+/// encoding of thread identity vended by libdispatch called dispatch_lock_t,
+/// and a futex-style dispatch API in order to escalate the priority of a thread.
+/// When the thread gives up running a task, the priority escalation that may
+/// have been applied on the thread, is removed. Henceforth, the dispatch_lock_t
+/// tracked in the ActiveTaskStatus will be called the ExecutionLock.
+///
+/// In order to keep the entire ActiveTaskStatus size to 2 words, the thread
+/// identity is only tracked on platforms which can support 128 bit atomic
+/// operations. The ActiveTaskStatus's layout has thus been changed to have the
+/// following layout depending on the system configuration supported:
+///
+/// 32 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=1
+///
+///          Flags               Exeuction Lock           Unused           TaskStatusRecord *
+/// |----------------------|----------------------|----------------------|-------------------|
+///          32 bits                32 bits                32 bits              32 bits
+///
+/// 64 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=1
+///
+///         Flags              Execution Lock      TaskStatusRecord *
+/// |----------------------|-------------------|----------------------|
+///          32 bits                32 bits              64 bits
+///
+/// 32 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=0
+///
+///          Flags             TaskStatusRecord *
+/// |----------------------|----------------------|
+///          32 bits                32 bits
+//
+/// 64 bit systems with SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION=0
+///
+///         Flags                  Unused            TaskStatusRecord *
+/// |----------------------|----------------------|---------------------|
+///         32 bits                 32 bits                64 bits
+///
+/// Note: This layout is intentional so that we can then do 64 bit atomics on
+/// just the top 64 bits of atomic state, including escalation of a thread. We
+/// only need to do double wide atomics if we need to reach for the
+/// StatusRecord pointers and therefore have to update the flags at the same
+/// time.
```

Comment:
Thanks, this comment looks great.

Well, except this last paragraph is confusing.  We're not actually doing partial atomic updates, right?  It's just the partial atomic read from the kernel.

Reply:
Currently we are doing full double wide atomic updates always but we don't always need to do that.

We can do relaxed loads on the top word and only do the actual double-wide atomics if the isStatusRecordLockedBit is set and we care to take the record lock. If we don't need the lock (like for instance, for cancellation of a task that has no children), then we don't need to double-wide atomics. It's a nice micro-optimization that is available to us and will ease register pressure in a lot of points.